### PR TITLE
[web][font] Make expo-font more agnostic on web

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

It should be much easier to use this library by itself in the browser. Currently using **expo-font** requires **expo-constants**, **expo-asset**, **expo-file-system**, **expo-permissions**, and all of the Unimodule core libraries.

# How

- Split the loading logic into different files
- Prevent Asset, and Constants from being used on web
- Reuse the same style element for fonts instead of creating a new one each time
- Added ability to define the `font-display` property
- Exporting new `FontDisplay` enum which defaults to `FontDisplay.Auto` on web and only allows for `FontDisplay.Auto` on native.

# Test Plan

- Update storybook to ensure the font-display is being added to the snapshots.

### CODE REVIEW

File: packages/expo-font/src/FontLoader.web.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import ExpoFontLoader from './ExpoFontLoader';
+import { canUseDOM } from 'fbjs/lib/ExecutionEnvironment';
+import { FontSource, FontResource } from './FontTypes.web';
+/**
+ * A font source can be a URI, a module ID, or an Expo Asset.
```
Comment (ide) [inline packages/expo-font/src/FontLoader.web.ts]: remove comment

File: packages/expo-font/src/FontLoader.web.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+export function fontFamilyNeedsScoping(name: string): boolean {
+  return false;
+}
+
+export function getAssetForSource(source: FontSource): FontResource {
+  if (typeof source === 'object' && 'uri' in source) {
+    return {
+        display: source.display,
+        // @ts-ignore
+        uri: source.uri || source.localUri
```
Comment (ide) [inline packages/expo-font/src/FontLoader.web.ts]: where does `localUri` come from?

File: packages/expo-font/src/FontTypes.web.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+
```
Comment (ide) [inline packages/expo-font/src/FontTypes.web.ts]: Platform-specific extensions like `.web.ts` don't work with types. The declared types need to work on all platforms simultaneously.

File: packages/expo-font/src/FontTypes.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import { Asset } from 'expo-asset';
+
+/**
+ * A font source can be a URI, a module ID, or an Expo Asset.
+ */
+export type FontSource = string | number | Asset | FontResource;
+
+// noop: web only
+export enum FontDisplay {
+    Auto = 'auto',
```
Comment (ide) [inline packages/expo-font/src/FontTypes.ts]:
```suggestion
    AUTO = 'auto',
```

File: packages/expo-font/src/FontTypes.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import { Asset } from 'expo-asset';
+
+/**
+ * A font source can be a URI, a module ID, or an Expo Asset.
+ */
+export type FontSource = string | number | Asset | FontResource;
+
+// noop: web only
```
Comment (ide) [inline packages/expo-font/src/FontTypes.ts]: document this more thoroughly, especially if the enum values must match web values (and link to the spec)

File: packages/expo-font/src/FontLoader.web.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    throw new Error(`Unexpected type ${typeof source} expected a URI string or Asset from expo-asset.`);
+  }
+
+  return {
+      uri: source
+    };
+}
+
+export async function loadSingleFontAsync(name: string, asset: FontResource): Promise<void> {
+    if (canUseDOM) {
```
Comment (ide) [inline packages/expo-font/src/FontLoader.web.ts]: is this check needed?

**Review by ide — CHANGES_REQUESTED** [review]
The fact that Asset is supported on some platforms but not others is fragile and goes against Expo. We need to make the API universal. If the goal is to have a way to use fonts on web without expo-asset,  the universal way to do this is to make a way to use fonts on all platforms without expo-asset.

File: packages/expo-font/src/FontLoader.web.ts (hunk 2 above)
Comment (EvanBacon) [reply packages/expo-font/src/FontLoader.web.ts]: if someone passes a loaded `Asset`

Comment (ide) [reply packages/expo-font/src/FontLoader.web.ts]: We then either need to add `localUri` to the FontSource type, and fix the `uri in source` check above, or remove this. Right now the program is not clear.

File: packages/expo-font/src/FontLoader.web.ts (hunk 6 above)
Comment (EvanBacon) [reply packages/expo-font/src/FontLoader.web.ts]: SSR

File: packages/expo-font/src/FontTypes.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+export enum FontDisplay {
+  AUTO = 'auto',
+  BLOCK = 'block',
+  SWAP = 'swap',
+  FALLBACK = 'fallback',
+  OPTIONAL = 'optional',
+}
+
+export interface FontResource {
```
Comment (ide) [inline packages/expo-font/src/FontTypes.ts]:
```suggestion
export type FontResource = {
```

**Review by ide — CHANGES_REQUESTED** [review]

**Review by ide — APPROVED** [review]

**Pull request merged**